src/bp/ui-studio/src/web/actions/index.ts

```diff
@@ -11,71 +16,114 @@ export const receiveFlows = createAction('FLOWS/RECEIVE', flows => flows, () =>
+export const requestDeleteFlow = createAction('FLOWS/DELETE')
+export const requestDuplicateFlow = createAction('FLOWS/DUPLICATE')
+
+export const requestUpdateFlowNode = createAction('FLOWS/FLOW/UPDATE_NODE')
+export const requestCreateFlowNode = createAction('FLOWS/FLOW/CREATE')
+export const requestRemoveFlowNode = createAction('FLOWS/FLOW/REMOVE')
+
+export const requestPasteFlowNode = createAction('FLOWS/NODE/PASTE')
+export const requestPasteFlowNodeElement = createAction('FLOWS/NODE_ELEMENT/PASTE')
+
+const wrapAction = (
+  requestAction,
+  asyncCallback: AsyncCallback,
+  receiveAction = receiveSaveFlows,
+  errorAction = errorSaveFlows
+) => payload => (dispatch, getState) => {
+  dispatch(requestAction(payload))
+  // tslint:disable-next-line: no-floating-promises
+  asyncCallback(payload, getState())
+    .then(() => dispatch(receiveAction()))
+    .catch(err => {
+      dispatch(errorAction(err))
+    })
+}
 
-export const saveAllFlows = () => (dispatch, getState) => {
-  dispatch(requestSaveFlows())
+const updateCurrentFlow = async (_payload, state) => {
+  const flowState = state.flows
+  await FlowsAPI.updateFlow(flowState, flowState.currentFlow)
+}
 
-  const state = getState()
-  const flowsByName = state.flows.flowsByName
-
-  let dirtyFlows = getDirtyFlows(state)
-  const cleanFlows = _.chain(flowsByName)
-    .values()
-    .map(flow => flow.name)
-    .difference(dirtyFlows)
-    .value()
-
-  dirtyFlows = dirtyFlows
-    .filter(name => !!flowsByName[name])
-    .map(name => {
-      const flow = flowsByName[name]
-      return {
-        name,
-        version: '0.0.1',
-        flow: name,
-        location: flow.location,
-        startNode: flow.startNode,
-        catchAll: flow.catchAll,
-        links: flow.links,
-        nodes: flow.nodes,
-        skillData: flow.skillData,
-        timeoutNode: flow.timeoutNode
-      }
-    })
+const saveDirtyFlows = async flowState => {
+  const dirtyFlows = getModifiedFlows(flowState).filter(name => !!flowState.flowsByName[name])
 
-  axios.post(`${window.BOT_API_PATH}/flows`, { cleanFlows, dirtyFlows }).then(() => {
-    dispatch(receiveSaveFlows())
-  })
+  for (const flow of dirtyFlows) {
+    await FlowsAPI.updateFlow(flowState, flow)
+  }
 }
 
-export const updateFlow = createAction('FLOWS/FLOW/UPDATE')
-export const renameFlow = createAction('FLOWS/FLOW/RENAME')
-export const createFlow = createAction('FLOWS/CREATE')
+export const updateFlow = wrapAction(requestUpdateFlow, updateCurrentFlow)
+
+export const renameFlow = wrapAction(requestRenameFlow, async (payload, state) => {
+  const { targetFlow, name } = payload
+  const flowState = state.flows
+  await FlowsAPI.renameFlow(flowState, targetFlow, name)
+  await saveDirtyFlows(flowState)
+})
+
+export const createFlow = wrapAction(requestCreateFlow, async (payload, state) => {
+  const name = payload
+  const flowState = state.flows
+  await FlowsAPI.createFlow(flowState, name)
```

Comment:
There's too much boilerplate and moving between files to understand the logic. I've only refactored it for createFlow, but don't you think this would be clearer? 

```js
export const createFlow = wrapAction(requestCreateFlow, async (name, state) => {
 await debounceInsertFlow(toFlowDto(state.flows.flowsByName[name], name))
})
```

Reply:
I'm not against merging current api logic in one file, but I don't think API code should be merged with actions code.

I think it's important to correctly separate concerns.